1372: Add suffixes for matching glossary terms

pedro-mendonca opened the pull request:
Add suffixes to matching glossary terms.

# Nouns
https://www.thefreedictionary.com/Forming-Plurals.htm
https://preply.com/en/blog/simple-rules-for-the-formation-of-plural-nouns-in-english/

### Plurals of singular nouns
- [x] Ending in a sibilant (-ss, -z, -x, -sh, -ch). **Suffix: '-es'.**
      *E.g.: Kiss and kiss-es, waltz and waltz-es, box and box-es, dish and dish-es, coach and coach-es.*
- [x] Ending with '-y' preceded by vowel. **Suffix: '-s'.**
      *E.g.: Delay and delay-s, key and key-s, toy and toy-s, guy and guy-s.*
- [x] Ending with '-o' and '-y' preceded by consonant. **Suffix: '-es'.**
      *E.g.: Lady and ladi-es, hero, tomato.*
- [x] Ending with with '-an'. **Suffix: '-en'.**
      *E.g.: Woman and wom-en.*
- [x] Ending with '-f', '-fe' or '-s'. **Suffix: '-es'.**
      *E.g.: Wife and wiv-es, leaf and leav-es, wolf and wolv-es, bus and bus-es.*
- [x] Fallback suffix for most nouns not ended with '-s'. **Suffix: '-s'.**

# Verbs

### Third-person singular
- [x] Ending in a sibilant (-ss, -z, -x, -sh, -ch). **Suffix: '-es'.**
      *E.g.: Pass and pass-es, quiz and quiz-es, fix and fix-es, push and push-es, watch and watch-es.*
- [x] Ending with '-y' preceded by vowel. **Suffix: '-s'.**
      *E.g.: Play and play-s.*
- [x] Ending with '-o' and '-y' preceded by consonant. **Suffix: '-es'.**
      *E.g.: Try and tri-es, Go and go-es, do and do-es.*
- [x] Fallback suffix for most verbs. **Suffix: '-s'.**
      *E.g.: Format and format-s, make and make-s, pull and pull-s*

### Past simple tense and past participle. **Suffix '-ed'.**
- [x] Not ending with '-e'.
      *E.g.: Fix and fix-ed, push and push-ed.*
- [x] Ending with '-e'.
      *E.g.: Contribute and contribut-ed, delete and delet-ed.*

### Present participle and gerund. **Suffix '-ing'.**
- [x] Not ending with '-e'.
      *E.g.: Fix and fix-ing, push and push-ing.*
- [x] Ending with '-ee', '-ye' or '-oe'.
      *E.g.: Agree and agree-ing, see and see-ing, dye and dye-ing, tiptoe and tiptoe-ing.*
- [x] Ending with single '-e'.
      *E.g.: Contribute and contribut-ing, delete and delet-ing, care and car-ing. Change to '-ing'.*

### Nouns formed by Verbs
https://www.thefreedictionary.com/Commonly-Confused-Suffixes-tion-vs-sion.htm
- [x] Verbs that form nouns ending with **Suffix '-tion'**.
      *Check the complete list below.*
```
// General.
'ate'    => 'a',     // Abbreviate and abbrevia-tion. Change to 'a-tion'.
'ize'    => 'iza',   // Authorize and authoriza-tion. Change to 'iza-tion'.
'ify'    => 'ifica', // Specify and specifica-tion. Change to 'ifica-tion'.
'efy'    => 'efac',  // Liquefy and liquefac-tion. Change to 'efaca-tion'.
'aim'    => 'ama',   // Exclaim and exclama-tion. Change to 'ama-tion'.
'pt'     => 'p',     // Encrypt and encryp-tion. Change to 'p-tion'.
'scribe' => 'scrip', // Subscribe and subscrip-tion. Change to 'scrip-tion'.
'ceive'  => 'cep',   // Perceive and percep-tion. Change to 'cep-tion'.
'sume'   => 'sump',  // Resume and resump-tion. Change to 'sump-tion'.
'ct'     => 'c',     // Correct and correc-tion. Change to 'c-tion'.
'ete'    => 'e',     // Delete and dele-tion. Change to 'e-tion'.
'it'     => 'i',     // Edit and edi-tion. Change to 'i-tion'.
'ite'    => 'i',     // Ignite and igni-tion. Change to 'i-tion'.
'ute'    => 'u',     // Contribute and contribu-tion. Change to 'u-tion'.
'olve'   => 'olu',   // Resolve and resolu-tion. Change to 'olu-tion'.
'ose'    => 'osi',   // Compose and composi-tion. Change to 'osi-tion'.
// After 'n' cases.
'tain'   => 'ten',   // Abstain and absten-tion. Change to 'ten-tion'.
'vene'   => 'ven',   // Contravene and contraven-tion. Change to 'ven-tion'.
'vent'   => 'ven',   // Prevent and preven-tion. Change to 'ven-tion'.
// After 'r' cases.
'rt'     => 'r',     // Insert and inser-tion. Change to 'r-tion'.
```
- [x] Verbs that form nouns ending with **Suffix '-sion'**.
      *Check the complete list below.*
```
// General.
'ade'  => 'a',   // Invade and inva-sion. Change to 'a-sion'.
'cede' => 'ces', // Precede and preces-sion. Change to 'ces-sion'.
'ide'  => 'i',   // Decide and deci-sion. Change to 'i-sion'.
'ode'  => 'o',   // Explode and explo-sion. Change to 'o-sion'.
'ude'  => 'u',   // Exclude and exclu-sion. Change to 'u-sion'.
'ise'  => 'i',   // Supervise and supervi-sion. Change to 'i-sion'.
'use'  => 'u',   // Confuse and confu-sion. Change to 'u-sion'.
'pel'  => 'pul', // Expel and expul-sion. Change to 'pul-sion'.
'mit'  => 'mis', // Submit and submis-sion. Change to 'mis-sion'.
'ss'   => 's',   // Compress and compres-sion. Change to 's-sion'.
// After 'n' cases.
'end'  => 'en',  // Extend and exten-sion. Change to 'en-sion'.
// After 'r' cases.
'vert' => 'ver', // Convert and conver-sion. Change to 'ver-sion'.
'erse' => 'er',  // Disperse and disper-sion. Change to 'er-sion'.
'ur'   => 'ur',  // Recur and recur-sion. Change to 'ur-sion'.
'erge' => 'er',  // Emerge and emer-sion. Change to 'er-sion'.
```

# Hooks
- [x] Add filter `gp_glossary_match_suffixes()` to allow customization.

Fixes #1384, #1386 and #1372

pedro-mendonca committed:
Add 'ed', 'ion', and 'ing' to suffix list for glossary term ending with 'e'

ocean90 commented on gp-templates/helper-functions.php:
```diff
@@ -108,6 +108,10 @@ function gp_sort_glossary_entries_terms( $glossary_entries ) {
 			$terms[] = substr( $term, 0, -2 ) . 'ves';
+		} elseif ( 'e' === substr( $term, -1 ) ) {
+			$terms[] = substr( $term, 0, -1 ) . 'ed';
+			$terms[] = substr( $term, 0, -1 ) . 'ion';
```
I think adding `ion` should only be added for specific last characters (after removing `e`).  Some examples where `ion` doesn't make sense: `wave`, `house`, `care`, `sensitive`

pedro-mendonca replied:
Could it be the case only for words ending with `te`?

pedro-mendonca replied:
Here is a comprehensive list of suffix rules for -ion
https://www.thefreedictionary.com/Commonly-Confused-Suffixes-tion-vs-sion.htm

There are many other cases, I'll add this to the issue so it can be discussed also.

pedro-mendonca committed:
Add known suffixes for nouns and verbs

pedro-mendonca commented:
@ocean90 I've made a [big change](https://github.com/GlotPress/GlotPress/pull/1373/commits/72fffd6fc4c47f08d3e16266ef84a235c82b187e) by separating terms by type, with more clear lists that we can keep or comment out, for testing.

Let me know your thoughts on this please.

ocean90 commented:
I like the direction where this goes! But please make sure you update your branch with the latest changes from #1395 as it changes how `gp_sort_glossary_entries_terms()`, now `gp_glossary_add_suffixes()`, works.

ocean90 commented:
@pedro-mendonca Do you plan to continue working on this? Do you need any help?

pedro-mendonca commented:
Hi, I do, just had no time yet. I'll get back to it.
Thanks!

pedro-mendonca committed:
Revert previous commits to further rebase

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes-fix

pedro-mendonca committed:
Add known suffixes for nouns and verbs

pedro-mendonca committed:
Replace the whole list of known plurals by part_of_speech.

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca commented:
@ocean90 I have a .po with a set of originals and a .csv for the glossary to test this.
I can share here if you want.
I'll do some unit tests for all these cases.

pedro-mendonca commented:
After looking at most rules, I've made an array of rules with the same structure:
```
array(
  'endings'  => array(
    'y',  // Lady.
  ),
  'preceded' => '[b-df-hj-np-tv-xz]',  // Preceded by any consonant.
  'change'   => 'i',                   // Change to 'i'.
  'add'      => 'es',                  // Add 'es'.
),
```

Then I made a single logic that runs on every set of rules, for every `part_of_speech` type.

We can add rules or remove them by adding or removing these arrays.

Feedback please :)

pedro-mendonca commented:
@ocean90 I see two possible places where to put the unit tests:
https://github.com/GlotPress/GlotPress/blob/develop/tests/phpunit/testcases/test_template_helper_functions.php
and
https://github.com/GlotPress/GlotPress/blob/develop/tests/phpunit/testcases/tests_things/test_thing_glossary_entry.php
Which one should we go for?

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Only trims ending pipe

pedro-mendonca committed:
Check if key term is set and is an array

pedro-mendonca committed:
Make suffix doc more clear

pedro-mendonca committed:
Add unit tests for suffixes

pedro-mendonca commented:
@ocean90 I've added all the Unit Tests for the suffixes, for nouns and verbs.
Feel free to review when you can.

pedro-mendonca committed:
Fix tests descriptions

pedro-mendonca committed:
Rename variables for a generic approach for nouns and verbs

pedro-mendonca committed:
Improve list of Nouns formed by verbs with the suffix -ion

pedro-mendonca committed:
Move 'changes' to the pairs ending => change

pedro-mendonca committed:
Improve suffixes changes documentation

pedro-mendonca committed:
Test verbs that form nouns ending with -tion

pedro-mendonca committed:
Test verbs that form nouns ending with -sion

pedro-mendonca committed:
Remove unnecessary code

pedro-mendonca commented:
Added the missing tests for Verbs that form Nouns (suffixes -tion and -sion).
Improved the documention about the changes made in each word case ending.
Also changed a bit the structure in the suffix cases arrays to make it more generic and reusable.

The PR now includes a comprehensive list of suffixes for nouns and verbs.
We can choose which cases to keep or to comment out.

I think it's now ready for review. Sorry for the last day changes, many tests were missing,

pedro-mendonca committed:
Fix endings documentation and some formatting

pedro-mendonca commented:
@ocean90, @akirk and @amieiro I've updated the PR description above with a complete list of proposed changes, with the list of all Nouns and Verbs suffixes.

pedro-mendonca committed:
Apply coding standards

ocean90 commented:
Does this also take care of #1386 and/or #1384?

I think some before/after comparison with some real glossaries used on translate.w.org (like the one from #1377)  would be interesting. Do we have less or more terms? If less, are there any that should still be in the list?

pedro-mendonca committed:
Add link to the plurals formation documentation

pedro-mendonca commented:
> Does this also take care of https://github.com/GlotPress/GlotPress/issues/1386 and/or https://github.com/GlotPress/GlotPress/issues/1384?

Yes, it does, now the suffixes are only applied for each matching case.
Example bellow:
 - Only add suffix `-es` to words ending with `-o` and `-y`, preceded by any consonant.
 - For `-o` case, just add `-es`.
 - For `-y`, change `-y` to `-i` before add `-es`.
```
// Ending with '-o' and '-y' preceded by consonant. Suffix: '-es'.
array(
	'endings'  => array(
		'y' => 'i',  // Lady and ladi-es. Change to 'i-es'.
		'o' => null, // Hero and hero-es, tomato and tomato-es.
	),
	'preceded' => '[b-df-hj-np-tv-xz]', // Preceded by any consonant.
	'add'      => 'es',                 // Add 'es'.
),
```

> I think some before/after comparison with some real glossaries used on translate.w.org (like the one from https://github.com/GlotPress/GlotPress/issues/1377) would be interesting. Do we have less or more terms? If less, are there any that should still be in the list?

I can add some performance tests.

For now, besides the unit testing already added to this PR, I've made an [originals .pot file](https://github.com/pedro-mendonca/GlotPress/blob/glossary-suffix-testing-example/tests/phpunit/data/glossary-suffix-testing-example--originals.pot) and a [glossary CSV](https://github.com/pedro-mendonca/GlotPress/blob/glossary-suffix-testing-example/tests/phpunit/data/glossary-suffix-testing-example--locale-glossary.csv) with the examples in the unit tests, so it can be tested more visually.

pedro-mendonca commented:
> Do we have less or more terms? If less, are there any that should still be in the list?

Previously we had many unnecessary terms matching applied agnostically to any `part_of_speech` type.

Now we have specific rules for each `part_of_speech`. Currently only for Nouns and Verbs.
Each rule checks for the term ending, and also check what is preceeding that ending.
This granularity provides a much better control over what is really to be matched, with no (or minimal) unnecessary matching.

So, it's hard to say if there are more or less.
I'd say probably more terms because there are more rules, but less wrongly matched terms.

> ...are there any that should still be in the list?

I've compiled these lists of rules based on the links below, also added to the code as comments, if there are any others that need to be added, we just need to add it to the rules array.
https://preply.com/en/blog/simple-rules-for-the-formation-of-plural-nouns-in-english/
https://www.thefreedictionary.com/Forming-Plurals.htm.
https://www.thefreedictionary.com/Commonly-Confused-Suffixes-tion-vs-sion.htm.

pedro-mendonca committed:
Improve suffix comment

pedro-mendonca commented:
And now for some testing data.

### Data
Project: [WordPress core 6.0 - Dev](https://translate.wordpress.org/projects/wp/dev/): **5.356** originals.
Portuguese [Locale Glossary](https://translate.wordpress.org/locale/pt/default/glossary/): **277** entries.
Settings: **All the strings shown in a single page**.
PHP **7.4.1**

### Results
Page Generation Time: **23s** average both before and after this PR (data from Query Monitor)

There isn't much of a difference in performance, this PR doesn't change the logic/performance already set before for matching, only improves the suffixes rules organization, per ending, and per part_of_speech.

pedro-mendonca commented:
In an example from #1383 many errors ocur from glossary verbs being inserted in other forms than the infinitive, and from nouns being inserted in plural form, making it harder to process and predict. This PR doesn't change this, the verbs should be in the infinitive form and the nouns in singular, as you only have to match from the infinitive to other tenses and nouns from singular to plural. Many of these matches are many->one so it's not possible to match it reverse.

Some of the problematic examples mentioned in #1383:
```
'Customizing|Customizings|Customizinges|Customizinged|Customizinging'
'recommended|recommendeds|recommendedes|recommendeded|recommendeding'
```
The problem here is that the verbs aren't in the infinitive form.

With this PR, if both verbs are on infinitive, the result is:
```
customiz ( e | ed | e | ing )
custom ( ize | ization )
recommend ( s | ed | ing )
```
There is a redundancy still, `customiz-e` (2x) + `custom-ize`.
I think the 2x `customiz-e` is worth to remove.

pedro-mendonca committed:
Don't duplicate endings to glossary entry suffixes.

pedro-mendonca commented:
> `customiz ( e | ed | e | ing )`
> There is a redundancy still, `customiz-e` (2x) + `custom-ize`.

The commit https://github.com/GlotPress/GlotPress/pull/1373/commits/39d057f59b3f7f8b4afc3aee9ab06200385dbdff fixes the ending duplicates added to the array.
Now: `customiz ( e | ed | ing )`

pedro-mendonca commented:
> I think some before/after comparison with some real glossaries used on translate.w.org (like the one from https://github.com/GlotPress/GlotPress/issues/1377) would be interesting.

@ocean90 here is some testing data with the translate.w.org `kmr` Locale Glossary (the one from https://github.com/GlotPress/GlotPress/issues/1377)

### Data

Project: [WordPress core 6.0 - Dev](https://translate.wordpress.org/projects/wp/dev/): **5.356** originals.
Kurdish [Locale Glossary](https://translate.wordpress.org/locale/kmr/default/glossary/): **645** entries.
Settings: **All the strings shown in a single page**.
PHP **7.4.1**

### Results

Page Generation Time: **23s** average both before and after this PR (data from Query Monitor)
Same amount of time than the [previous test](https://github.com/GlotPress/GlotPress/pull/1373#issuecomment-1191513148) with **277** glossary entries (Portuguese from translate.w.org)

For comparison, the Page Generation Time for the same project (5.356 strings) on a translation set with an empty glossary is about **22s**.

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge remote-tracking branch 'upstream/develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca commented:
Can someone please have a look at this PR and give some feedback? :)

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

amieiro committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Combine the suffixes for shorter regular expression. (#1651)

* Combine the suffixes for shorter regular expression.

* Whitespace

* Add unit tests

---------

Co-authored-by: Alex Kirk <akirk@users.noreply.github.com>
# Conflicts:
#	tests/phpunit/testcases/test_template_helper_functions.php

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

# Conflicts:
#	tests/phpunit/testcases/test_template_helper_functions.php

pedro-mendonca committed:
We don't need it anymore because the match checks what precedes the ending

pedro-mendonca committed:
Fix part of speech for its own set of endings

pedro-mendonca committed:
Make the Suffix list filterable.

pedro-mendonca commented:
Hi team,

After merging the develop branch, found the following:
The commit https://github.com/GlotPress/GlotPress/commit/d93b07dd847187d47c083988cb424b2177ae01ad includes a test "I know, partys is the wrong plural ending but we need it because of nouns like boys." that doesn't match anymore because with this PR the endings aren't applied to all words, the part of speech must match, and what preceeds the ending is also checked.
The mentioned commit also have two matches for "report" that are verbs and not nouns, with this PR, for these matchs happen, fixed the test from 'noun' to 'verb'.

After fixing all this, both the PR tests and the previous are all ok again.

I've [added a filter](https://github.com/GlotPress/GlotPress/pull/1373/commits/42bd44c859cebcfabf499d47103e468bde50e84a) to allow the customization of this Suffixes list.

Please @ocean90 and @amieiro let me know if you have any other suggestions.

pedro-mendonca committed:
Rename filter and improve description

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Fix typo in tests part_of_speech introduced in PR #1706

amieiro reviewed (changes requested):
I have tested this PR and I give here my feedback. Congratulations, Pedro, this work is amazing!!!

We have these different types of elements in a glossary:
- Noun
- Verb
- Adjective
- Adverb
- Interjection
- Conjunction
- Preposition
- Pronoun
- Expression
- Abbreviation

This PR only checks the nouns and verbs, but we should check (I suggest in a new PR):
- Adjectives. E.g. happy -> happier.
- Pronouns. E.g., he -> she, him -> her.

I have used this documentation as base for this analysis:
- Nouns
  - https://www.thefreedictionary.com/Forming-Plurals.htm
  - https://preply.com/en/blog/simple-rules-for-the-formation-of-plural-nouns-in-english/
- Verbs
  - https://dictionary.cambridge.org/grammar/british-grammar/verbs-basic-forms
  - https://www.thefreedictionary.com/Inflection-in-Spelling.htm
- Nouns formed from verbs
  - https://www.thefreedictionary.com/Commonly-Confused-Suffixes-tion-vs-sion.htm
- 

I have used these files to test:
  - Pot file: https://github.com/pedro-mendonca/GlotPress/blob/glossary-suffix-testing-example/tests/phpunit/data/glossary-suffix-testing-example--originals.pot
  - Glossary file (CSV): https://github.com/pedro-mendonca/GlotPress/blob/glossary-suffix-testing-example/tests/phpunit/data/glossary-suffix-testing-example--locale-glossary.csv
 
## Nouns
- 1 ✅ Ending in "-ss," "-z," "-x," "-sh," "-ch," or "-tch." -> Add "es". "-tch" considered the same than "-ch".
  - Visual examples:
      - Kiss -> kisses. 
      - waltz -> waltzes.
      - box -> boxes.
      - dish -> dishes.
      - coach -> coaches.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_nouns_ending_with_sibilant_in_glossary (Test 1)
- 2 ✅ Ending in "-y" preceded by a vowel -> Add "s".
  - Visual examples:
      - Delay -> delays.
      - key -> keys.
      - toy -> toys. 
      - guy -> guys.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_nouns_ending_with_y_preceded_by_vowel_in_glossary (Test 2)
- 3 ✅ Ending in "-y" preceded by a consonant -> Change "y" to "i" + "es". Ending in consonant + "o" -> Add "es".
  - Visual examples:
      - Lady -> ladies.
      - Hero -> heroes. 
      - tomato -> tomatoes.
  - Unit test: 
      - test_map_glossary_entries_to_translation_originals_with_nouns_ending_with_y_preceded_by_consonant_in_glossary (Test 3)
      - test_map_glossary_entries_to_translation_originals_with_nouns_alternate_ending_with_o_preceded_by_consonant_in_glossary (Test 4)
- 4 ❌ Incorrect. Ending with '-an'. Suffix: '-en'. Words ending with 'an" add a "s" to make the plural. Some examples:
  - Plan -> Plans, Can -> Cans, Pan -> Pans, Fan -> Fans, Clan -> Clans, Span -> Spans, Human -> Humans.
  - In English we have some words with irregular plurals (Woman -> Women, Man -> Men, Child -> Children, Foot -> Feet, Tooth -> Teeth). I think it doesn't make sense to add a full list of irregular plurals.
  - We should remove the unit test test_map_glossary_entries_to_translation_originals_with_nouns_ending_with_an_in_glossary (Test 5)
- 5 ✅ Ending in "-f," "-fe" -> replaces the endings with "-v" and adds "es". Ending in "-s" -> adds "es"
  - Visual examples:
      - Wife -> wives.
      - Leaf -> leaves.
      - wolf -> wolves.
      - Bus -> buses.
      - lens -> lenses.
  - Unit test: 
    - test_map_glossary_entries_to_translation_originals_with_nouns_ending_with_f_in_glossary (Test 6)
    - test_map_glossary_entries_to_translation_originals_with_nouns_fallback_ending_with_s_in_glossary (Test 7)
- 6 ❌ Incorrect. Fallback. We have this regex: "\w(?<!z|x|sh|ch|s|y|an|fe)". We need to remove the "an" (point 4) and add the "f" (point 5). This is the proposed regex: "\w(?<!z|x|sh|ch|s|y|f|fe)". Adds "s".
  - Visual examples:
      - chef -> chefs.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_nouns_fallback_not_ending_with_s_in_glossary (Test 8)
 
❓Comment. Have you considered the plural of some "-s" and "-z" endings? For some nouns that end in "-s" or "-z", you have to double the "-s" or "-z" and add "-es". For example: 
  - fez –> fezzes
  - gas –> gasses
As it is not a rule for all endings, we can avoid it.

# Verbs
We don't consider the irregular verbs.
- 1 ❌ Incorrect. Third-person singular. Ending in "-ss," "-z," "-x," "-sh," "-ch," or "-tch." -> Add "es". "-tch" considered the same than "-ch". We need to add the "-s" situation (e.g. bias->biases).
  - Visual examples:
      - Pass -> passes.
      - Quiz -> quizes. (The correct plural is quizzes).
      - Fix -> fixes.
      - Push -> pushes.
      - Watch -> watches.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_verbs_3rdperson_ending_with_sibilant_in_glossary (Test 9)
  - Example to add:
    - focus -> focuses.
- 2 ✅ Third-person singular. Ending in "-y" preceded by a vowel -> Add "s".
  - Visual examples:
      - Play -> plays.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_verbs_3rdperson_ending_with_y_preceded_by_vowel_in_glossary (Test 10)
- 3 ✅ Third-person singular. Ending in "-y" preceded by a consonant -> Change "y" to "i" + "es". Ending in consonant + "o" -> Add "es".
  - Visual examples:
      - Try -> tries.
      - Go -> goes.
      - do -> does.
  - Unit test: 
    - test_map_glossary_entries_to_translation_originals_with_verbs_3rdperson_ending_with_y_preceded_by_consonant_in_glossary (Test 11)
    - test_map_glossary_entries_to_translation_originals_with_verbs_3rdperson_ending_with_o_preceded_by_consonant_in_glossary (Test 12)
- 4 ✅ Fallback. We have this regex: "\w(?<!z|x|sh|ch|s|y|o)". Adds "s".
  - Visual examples:
      - Format -> formats.
      - make -> makes.
      - pull -> pulls.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_most_verbs_3rdperson_fallback_in_glossary (Test 13)
- 5 ✅ Past simple tense and past participle of verbs. Suffix '-ed'. Ending in "-e" -> Add "d". Other situations -> Add "ed". 
  - Visual examples:
    - Fix -> fixed.
    - push -> pushed.
    - Contribute -> contributed.
    - delete -> deleted.
  - Unit test: 
    - test_map_glossary_entries_to_translation_originals_with_verbs_past_ending_with_e_in_glossary (Test 14)
    - test_map_glossary_entries_to_translation_originals_with_verbs_past_not_ending_with_e_in_glossary (Test 15)
- 6 ✅ Present participle and gerund of verbs. Not ending with '-e', or ending with '-ee', '-ye' -or '-oe' -> Add "ing".
  - Visual examples:
    - Fix -> fixing.
    - push -> pushing.
    - Agree -> agreeing.
    - see -> seeing.
    - Dye -> dyeing.
    - Tiptoe -> tiptoeing.
    - Contribute -> contributing.
    - delete -> deleting.
    - care -> caring.
  - Unit test: 
    - test_map_glossary_entries_to_translation_originals_with_verbs_present_ending_with_e_in_glossary (Test 16)
    - test_map_glossary_entries_to_translation_originals_with_verbs_present_not_ending_with_e_or_ending_with_ee_ye_oe_in_glossary (Test 17)
- 7 ✅ Nouns formed from verbs ending in "tion".
  - We have different endings with different pattern substitutions who adds 'tion' after the next substitution.
    - 'ate'    -> 'a'.    
    - 'ize'    -> 'iza'.  
    - 'ify'    -> 'ifica'.
    - 'efy'    -> 'efac'. 
    - 'aim'    -> 'ama'.  
    - 'pt'     -> 'p'.    
    - 'scribe' -> 'scrip'.
    - 'ceive'  -> 'cep'.  
    - 'sume'   -> 'sump'. 
    - 'ct'     -> 'c'.    
    - 'ete'    -> 'e'.    
    - 'it'     -> 'i'.    
    - 'ite'    -> 'i'.    
    - 'ute'    -> 'u'.    
    - 'olve'   -> 'olu'.  
    - 'ose'    -> 'osi'.  
    - 'tain'   -> 'ten'.  
    - 'vene'   -> 'ven'.  
    - 'vent'   -> 'ven'.  
    - 'rt'     -> 'r'.    
  - Visual examples: 
    - Abbreviate -> abbreviation.
    - Authorize -> authorization.
    - Specify -> specification.
    - Liquefy -> liquefaction.
    - Exclaim -> exclamation.
    - Encrypt -> encryption.
    - Subscribe -> subscription. 
    - Perceive -> perception.
    - Resume -> resumption.
    - Correct -> correction.
    - Delete -> deletion. 
    - Edit -> edition. 
    - Ignite -> ignition. 
    - Contribute -> contribution.
    - Resolve -> resolution.
    - Compose -> composition.
    - Abstain -> abstention.
    - Contravene -> contravention
    - Prevent -> prevention.
    - Insert -> insertion.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_verbs_form_nouns_ending_with_tion_in_glossary (Test 18)
- 8 ✅ Nouns formed from verbs ending in "sion".
  - We have different endings with different pattern substitutions who adds 'sion' after the next substitution.
  - Visual examples:
    - Invade -> invasion. 
    - Precede -> precession. 
    - Decide -> decision.
    - Explode -> explosion.
    - Exclude -> exclusion.
    - Supervise -> supervision.
    - Confuse -> confusion.
    - Expel -> expulsion.
    - Submit -> submission.
    - Compress -> compression. 
    - Extend -> extension.
    - Convert -> conversion.
    - Disperse -> dispersion.
    - Recur -> recursion.
    - Emerge -> emersion.
  - Unit test: test_map_glossary_entries_to_translation_originals_with_verbs_form_nouns_ending_with_sion_in_glossary (Test 19)

❓Comment. [Here](https://dictionary.cambridge.org/grammar/british-grammar/verbs-basic-forms) we have some examples not checked:
- Base forms which end in vowel + single consonant. This depends on the stressed syllable, so we can create both, as we won't be able to know the stressed syllable. 
  - commit. -> committed, commited (incorrect), committing, commiting (incorrect)
- Vowel + l. The consonant is doubled if the base form ends in a vowel + l, whether the last syllable is stressed or not. I think we can't double the vowel with the current approach.
  - travel -> travelling, travelled

## Proposal to refactor the tests

As most of the test do the same checks with different words, I suggest to extract the repeated code to a new function and then use it in each test. The new function could be something like this:

```
	/**
	 * Method to test the map_glossary_entries_to_translation_originals() function.
	 *
	 * @param $test_string		  The string to test.
	 * @param $part_of_speech	  The part of speech of the string to test.
	 * @param $matches			    The matches to expect.
	 * @param $glossary_entries	The glossary entries to create.
	 * @return void
	 */
	function check_map_glossary( $test_string, $part_of_speech, $matches, $glossary_entries ) {
		$expected_result = array();
		foreach ( $matches as $glossary_entry => $originals ) {
			foreach ( $originals as $original ) {
				$expected_result[] = '<span class="glossary-word" data-translations="[{&quot;translation&quot;:&quot;' . $glossary_entry . '&quot;,&quot;pos&quot;:&quot;' . $part_of_speech . '&quot;,&quot;comment&quot;:null,&quot;locale_entry&quot;:&quot;&quot;}]">' . $original . '</span>';
			}
		}

		$expected_result = sprintf(
			'Testing words %s.',
			implode( ', ', $expected_result )
		);

		$entry = new Translation_Entry( array( 'singular' => $test_string, ) );

		$set = $this->factory->translation_set->create_with_project_and_locale();
		$glossary = GP::$glossary->create_and_select( array( 'translation_set_id' => $set->id ) );

		foreach ( $glossary_entries as $glossary_entry ) {
			$glossary_entry['glossary_id'] = $glossary->id;
			GP::$glossary_entry->create_and_select( $glossary_entry );
		}

		$orig = map_glossary_entries_to_translation_originals( $entry, $glossary );

		$this->assertEquals( $orig->singular_glossary_markup, $expected_result );
	}
```

The `test_map_glossary_entries_to_translation_originals_with_nouns_alternate_ending_with_o_preceded_by_consonant_in_glossary` could be refactored with only the elements that it wants to check by itself.

```
	/**
	 * Expects matching the plurals of Nouns alternate ending with '-o' preceded by consonant. Suffix: '-es'.
	 */
	function test_map_glossary_entries_to_translation_originals_with_nouns_alternate_ending_with_o_preceded_by_consonant_in_glossary() {
		$test_string = 'Testing words hero, heroes, tomato, tomatoes.';
		$part_of_speech = 'noun';

		$matches = array(
			'her\u00f3i' => array( // Portuguese.
				'hero',              // Singular.
				'heroes',            // Plural.
			),
			'tomate' => array( // Portuguese.
				'tomato',        // Singular.
				'tomatoes',      // Plural.
			),
		);

		$glossary_entries = array(
			array(
				'term' => 'hero', // Hero and heroes.
				'part_of_speech' => $part_of_speech,
				'translation' => 'herói', // Portuguese.
				'glossary_id' => null,
			),
			array(
				'term' => 'tomato', // Tomato and tomatoes.
				'part_of_speech' => $part_of_speech,
				'translation' => 'tomate', // Portuguese.
				'glossary_id' => null,
			),
		);

		$this->check_map_glossary( $test_string, $part_of_speech, $matches, $glossary_entries );
	}
```

## Load time

I have used this script in my local machine to test with this files:
  - Pot file used: https://github.com/pedro-mendonca/GlotPress/blob/glossary-suffix-testing-example/tests/phpunit/data/glossary-suffix-testing-example--originals.pot
  - Glossary file (CSV) used: https://github.com/pedro-mendonca/GlotPress/blob/glossary-suffix-testing-example/tests/phpunit/data/glossary-suffix-testing-example--locale-glossary.csv


```
#!/bin/bash

# Define the URL of the local web page
URL="https://glotpress.test/glotpress/projects/test/testing-glossary/aa/default/"

# Number of times to measure the load time
NUM_ITERATIONS=300

# Initialize a variable to store the total load time
TOTAL_TIME=0

# Loop through the specified number of iterations
for ((i=1; i<=$NUM_ITERATIONS; i++)); do
    # Use curl to fetch the URL and measure the time it took
    load_time=$(curl -s -o /dev/null -w "%{time_total}\n" "$URL")
    #load_time=$(curl -s -w "%{time_total}\n" "$URL")

    # Add the load time to the total
    TOTAL_TIME=$(echo "$TOTAL_TIME + $load_time" | bc)

    # Print the load time for this iteration
    echo "Load time for iteration $i: ${load_time} seconds"
done

# Calculate the mean load time in milliseconds
mean_time=$(echo "scale=3; ($TOTAL_TIME / $NUM_ITERATIONS) * 1000" | bc)

# Print the mean load time in milliseconds
echo "Mean load time: ${mean_time} milliseconds"
``` 

This is the output after 300 executions (1 script execution):

- Before: Mean load time: 110.000 milliseconds
- With this PR: Mean load time: 112.000 milliseconds

I have made another test with a real plugin and a real glossary:
- [WooCommerce with the Galician translation](https://translate.wordpress.org/projects/wp-plugins/woocommerce/stable/gl/default/export-translations/?filters%5Btranslated%5D=yes&filters%5Bstatus%5D=current).
- [Galician glossary](https://translate.wordpress.org/locale/gl/default/glossary/-export/).

This is the output after 300 executions (1 script execution) and 100 items per page (hardcoded):
- Before: Mean load time: 179.000 milliseconds
- With this PR: Mean load time: 165.000 milliseconds

So this PR doesn't add any performance issue.

amieiro commented:
I add the screenshots from the previous links with the variations rules, so if some link will break in the future, we have here the screenshots.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/b43b9858-1b7e-4575-a465-fdf191a048cd)

![image](https://github.com/GlotPress/GlotPress/assets/1667814/57635a25-47ae-42f4-9152-1b4dab6abd6e)

![image](https://github.com/GlotPress/GlotPress/assets/1667814/214bbaa3-d36c-4adf-a6d5-30b45d86dd47)

![image](https://github.com/GlotPress/GlotPress/assets/1667814/1267b70e-c52f-4e9e-8874-48544f0611dd)

![image](https://github.com/GlotPress/GlotPress/assets/1667814/9140a7ff-6f48-4ab7-8e95-ca67e02a92b5)

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Remove matching for irregular Nouns ending with '-an'

pedro-mendonca committed:
Improve legibility and consistency of dockblocks comments

pedro-mendonca committed:
Add support for verbs ending with single sibilant '-s' (bias)

pedro-mendonca committed:
Fix error in ending size

pedro-mendonca committed:
Improve glossary matches doc comments

pedro-mendonca committed:
Change matches array keys order, preceded first

pedro-mendonca committed:
Use actual preceded for any letters instead of null value.

pedro-mendonca committed:
Add support for placeholders in the endings, to allow reuse of the current ending

pedro-mendonca committed:
Add matches that double the ending consonant

pedro-mendonca committed:
Add check_map_glossary() to test different glossary matches

pedro-mendonca committed:
Fix comment.

pedro-mendonca committed:
Add all the glossary matches tests using check_map_glossary()

pedro-mendonca committed:
Remove irregular Noun plural match

pedro-mendonca committed:
Add test for verb focus

pedro-mendonca committed:
Fixed comments

pedro-mendonca commented:
@amieiro I've completed the full update related to your above suggestions.

### NOUNS

Here is a summary:
> 4 ❌ Incorrect. Ending with '-an'. Suffix: '-en'. Words ending with 'an" add a "s" to make the plural. Some examples:
>    Plan -> Plans, Can -> Cans, Pan -> Pans, Fan -> Fans, Clan -> Clans, Span -> Spans, Human -> Humans.
    In English we have some words with irregular plurals (Woman -> Women, Man -> Men, Child -> Children, Foot -> Feet, Tooth -> Teeth). I think it doesn't make sense to add a full list of irregular plurals.
    We should remove the unit test test_map_glossary_entries_to_translation_originals_with_nouns_ending_with_an_in_glossary (Test 5)

Fixed, removed all the entries and tests, old and new, about the irregular noun plurals for 'man' and 'woman'.

<hr>

> 6 ❌ Incorrect. Fallback. We have this regex: "\w(?<!z|x|sh|ch|s|y|an|fe)". We need to remove the "an" (point 4) and add the "f" (point 5). This is the proposed regex: "\w(?<!z|x|sh|ch|s|y|f|fe)". Adds "s".
>  Visual examples:
        chef -> chefs.
    Unit test: test_map_glossary_entries_to_translation_originals_with_nouns_fallback_not_ending_with_s_in_glossary (Test 8)

Fixed, removed the exception 'an'. No need to add 'f', as it's matched by the fallback default '-s'.

<hr>

>❓Comment. Have you considered the plural of some "-s" and "-z" endings? For some nouns that end in "-s" or "-z", you have to double the "-s" or "-z" and add "-es". For example:
>    fez –> fezzes
    gas –> gasses
    As it is not a rule for all endings, we can avoid it.

Not added, as it's not a regular case.

<hr>

### VERBS

> 1 ❌ Incorrect. Third-person singular. Ending in "-ss," "-z," "-x," "-sh," "-ch," or "-tch." -> Add "es". "-tch" considered the same than "-ch". We need to add the "-s" situation (e.g. bias->biases).
>    Visual examples:
        Pass -> passes.
        Quiz -> quizes. (The correct plural is quizzes).
        Fix -> fixes.
        Push -> pushes.
        Watch -> watches.
    Unit test: test_map_glossary_entries_to_translation_originals_with_verbs_3rdperson_ending_with_sibilant_in_glossary (Test 9)
    Example to add:
        focus -> focuses.

Fixed, removed the '-ss' and added '-s' to match both situations. Added case bias and focus.

<hr>

> ❓Comment. [Here](https://dictionary.cambridge.org/grammar/british-grammar/verbs-basic-forms) we have some examples not checked:
>    Base forms which end in vowel + single consonant. This depends on the stressed syllable, so we can create both, as we won't be able to know the stressed syllable.
        commit. -> committed, commited (incorrect), committing, commiting (incorrect)
    Vowel + l. The consonant is doubled if the base form ends in a vowel + l, whether the last syllable is stressed or not. I think we can't double the vowel with the current approach.
        travel -> travelling, travelled

Fixed: Now the ending values don't use null if no changes, now it has support for placeholders (%s or %1$s) to allow using the current ending, and duplicating for the case of doubling the ending consonants (committed, committing, etc.). The vowel-l case is matched in the added rules. As you've mentioned, it's not simple to detect the stressed syllable, so there are two matches for both single and double ending consonant.

<hr>

### TESTING
Created the function `check_map_glossary()` to simplify the testing for the functions `map_glossary_entries_to_translation_originals......()`.

<hr>

At last, do you think it would be usefull to add to the `tests/phpunit/data` a testing `.pot` and a glossary `.csv` despite not being used currently on any phpunit tests?

Thank you for your compreehensive review!

pedro-mendonca committed:
Add data files for manual testing

pedro-mendonca committed:
Add glossary matching case example in docblock

pedro-mendonca committed:
Remove ending newlines in example data file

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

pedro-mendonca committed:
Merge branch 'develop' into issue-suffixes

amieiro commented on tests/phpunit/testcases/test_template_helper_functions.php:
```diff
@@ -549,4 +538,1313 @@ function test_map_glossary_entries_to_translation_originals_with_suffixes( $test
+	/**
+	 * Method to test the map_glossary_entries_to_translation_originals() function.
+	 *
+	 * @param $test_string      The string to test.
+	 * @param $part_of_speech   The part of speech of the string to test.
+	 * @param $matches          The matches to expect.
+	 * @param $glossary_entries The glossary entries to create.
```
One small change in the header of the `check_map_glossary` method: I have added the type of each parameter, so the IDE doesn't show any warning and we can use it in the documentation and in another similar places: 

```suggestion
	 * @param $test_string      string The string to test.
	 * @param $part_of_speech   string The part of speech of the string to test.
	 * @param $matches          array  The matches to expect.
	 * @param $glossary_entries array  The glossary entries to create.
```

pedro-mendonca replied:
You're right, missed the types.
Still, it should be before the variable:
`@param string $var`

Right?

amieiro reviewed (changes requested):
Related with the .pot and .csv files in the `tests/phpunit/data`, as we don't use it in the automatic test, I suggest you to remove it from the PR and add it to a GitHub Gist or to a own repo. I think a good option is to add them in a GitHub Gist, so we can update it easily and discuss. But I think the best option will be to read both files in each test and using their data instead of hard-coding them (we can do this in a different PR).

pedro-mendonca committed:
Remove glossary suffix .pot and .csv example files

pedro-mendonca committed:
Add check_map_glossary() docblock types

pedro-mendonca commented:
> Related with the .pot and .csv files in the `tests/phpunit/data`, as we don't use it in the automatic test, I suggest you to remove it from the PR and add it to a GitHub Gist or to a own repo. I think a good option is to add them in a GitHub Gist, so we can update it easily and discuss. But I think the best option will be to read both files in each test and using their data instead of hard-coding them (we can do this in a different PR).

Hi @amieiro, thanks for your review!!
I've removed both glossary testing example files, .pot and .csv.
I've also added the missing types from check_map_glossary().
